rest-api/offline/public_metrics_export.py

```diff
@@ -0,0 +1,145 @@
+from sqlalchemy import text
+from dao import database_factory
+from dao.database_utils import replace_years_old
+from participant_enums import EnrollmentStatus, OrderStatus, PhysicalMeasurementsStatus
+from participant_enums import Race, QuestionnaireStatus
+
+# TODO(calbach): consider whether we need to filter down to consented
+# individuals only.
+
+def _questionnaire_metric(name, col):
+  """Returns a metrics SQL aggregation tuple for the given key/column."""
+  return (
+      name,
+      """
+      SELECT {}, SUM(1)
+      FROM participant_summary
+      GROUP BY 1;
+      """.format(col),
+      lambda v: QuestionnaireStatus.lookup_by_number(v).name
+  )
+
+# Metrics definitions. 3-tuples of:
+# - (str) aggregation key name
+# - (str) SQL statement to select value, count for a metric (in that order)
+# - (func(str): str) optional function which takes the value from the above SQL
+#   output and converts it for presentation
+_SQL_AGGREGATIONS = [
+  ('enrollmentStatus',
+   """
+   SELECT enrollment_status, SUM(1)
+   FROM participant_summary
+   GROUP BY 1;
+   """,
+   lambda v: EnrollmentStatus.lookup_by_number(v).name),
+  # TODO(calbach): Verify whether we need to be conditionally trimming these
+  # prefixes or leaving them unmodified. Unclear if all codes will have prefix
+  # "PMI_".
+  ('gender',
+   """
+   SELECT code.value, ps.count
+   FROM (
+    SELECT gender_identity_id, SUM(1) count
+    FROM participant_summary
+    WHERE gender_identity_id IS NOT NULL
+    GROUP BY 1
+   ) ps LEFT JOIN code
+   ON ps.gender_identity_id = code.code_id;
+   """,
+   None),
+  ('race',
+   """
+   SELECT race, SUM(1)
+   FROM participant_summary
+   WHERE race IS NOT NULL
+   GROUP BY 1
+   """,
+   lambda v: Race.lookup_by_number(v).name),
+  ('state',
+   """
+   SELECT SUBSTR(code.value, {}), ps.count
+   FROM (
+    SELECT state_id, SUM(1) count
+    FROM participant_summary
+    WHERE state_id IS NOT NULL
+    GROUP BY 1
+   ) ps LEFT JOIN code
+   ON ps.state_id = code.code_id;
+   """.format(len('PIIState_') + 1),
+   None),
+  ('ageRange',
+   """
+   SELECT
+     CASE
+      WHEN YEARS_OLD[date_of_birth] < 0 THEN 'UNSET'
```

Comment:
It's not obvious how YEARS_OLD(null) is returning < 0 here. Should we have a NULL check somewhere?

Also, I think some of these expressions return UNSET and others don't (e.g. state.) We should probably return UNSET everywhere, and make sure we have tests for it.

Reply:
I don't think we want to make the client work out counts by omission.

It should be OK to provide UNSET everywhere, and let the client decide whether or not to render it.

Reply:
According to the current spec they wouldn't have to do that since they aren't displaying them in the mock, but yeah - UNSET seems fine. Will update the design doc as well.